src/Aspire.Dashboard/Components/Controls/LogViewer.razor.cs

```diff
@@ -1,73 +1,74 @@
     {
-        if (_jsModule is not null)
+        if (firstRender)
         {
-            await _jsModule.InvokeVoidAsync("clearLogs", cancellationToken);
+            _jsModule ??= await JS.InvokeAsync<IJSObjectReference>("import", "/_content/Aspire.Dashboard/Components/Controls/LogViewer.razor.js");
+
+            _whenDomReady.TrySetResult();
         }
     }
 
-    private ValueTask WriteLogsToDomAsync(IEnumerable<LogEntry> logs)
-        => _jsModule is null ? ValueTask.CompletedTask : _jsModule.InvokeVoidAsync("addLogEntries", logs);
-
-    internal async Task WatchLogsAsync(Func<IAsyncEnumerable<string[]>> watchMethod, LogParserOptions logParserOptions = default)
+    internal async Task SetLogSourceAsync(IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>> batches, bool convertTimestampsFromUtc)
     {
-        var logParser = new LogParser(logParserOptions);
+        var cancellationToken = await _cts.NextAsync();
+        var logParser = new LogParser(convertTimestampsFromUtc);
 
-        await foreach (var logs in watchMethod())
+        // Ensure we are able to write to the DOM.
+        await _whenDomReady.Task;
+
+        await foreach (var batch in batches.WithCancellation(cancellationToken))
         {
-            var logEntries = new List<LogEntry>(logs.Length);
-            foreach (var log in logs)
+            if (batch.Count is 0)
             {
-                logEntries.Add(logParser.CreateLogEntry(log));
+                continue;
             }
 
-            await WriteLogsAsync(logEntries);
-        }
-    }
+            List<LogEntry> entries = new(batch.Count);
 
-    protected override async Task OnAfterRenderAsync(bool firstRender)
-    {
-        if (firstRender)
-        {
-            _jsModule ??= await JS.InvokeAsync<IJSObjectReference>("import", "/_content/Aspire.Dashboard/Components/Controls/LogViewer.razor.js");
-
-            while (_preRenderQueue.TryDequeue(out var logs))
+            foreach (var (content, isErrorOutput) in batch)
             {
-                await WriteLogsToDomAsync(logs);
+                entries.Add(logParser.CreateLogEntry(content, isErrorOutput));
             }
-            _renderComplete = true;
+
+            await _jsModule!.InvokeVoidAsync("addLogEntries", cancellationToken, entries);
         }
     }
 
-    private async Task WriteLogsAsync(IEnumerable<LogEntry> logs)
+    internal async Task ClearLogsAsync(CancellationToken cancellationToken = default)
     {
-        if (_renderComplete)
-        {
-            await WriteLogsToDomAsync(logs);
-        }
-        else
+        await _cts.ClearAsync();
+
+        if (_jsModule is not null)
         {
-            _preRenderQueue.Enqueue(logs);
+            await _jsModule.InvokeVoidAsync("clearLogs", cancellationToken);
         }
     }
 
     public async ValueTask DisposeAsync()
     {
+        _whenDomReady?.TrySetCanceled();
+
+        await _cts.ClearAsync();
+
         try
```

Comment:
Off topic: We should probably add a helper method for disposing modules and other JS types. It would centralize the try/catch error handling.

Reply:
Looks like there are three instances of this throughout the repo. I won't include that in this PR. To do it right you'd probably also want to ban the direct use of `IJSObjectReference.DisposeAsync`, which would mean setting up the banned API analyzer.